// porabote/comments/comments.js
import React, {Component} from 'react';
import {connect} from 'react-redux';
import Api from '@services/api-service';
import CommentForm from './comment-form';
import CommentList from './comment-list';
import './comments.less';

class Comments extends Component {

  state = {
    data: [],
    loading: true
  }

  componentDidMount() {
    this.fetchComments();
  }

  fetchComments = () => {

    this.setState({
      loding: true
    })

    Api.get(this.props.url, {
      query: {
        include: ['comments.user'],
        //orderBy: {parent_id: 'desc'}
      }
    }).then((data) => {
      this.setState({
        data: (typeof data.data !== 'undefined') ? data.data : [],
        loading: false
      })
    })
  }

  render() {

    return (
      <div className="comments">
        <CommentForm
          addUrl={this.props.addUrl}
          fetchRecord={this.fetchComments}
          auth={this.props.auth}
          {...this.props}
        />
        <CommentList
          auth={this.props.auth}
          data={this.state.data}
          fetchComments={this.fetchComments}
          loading={this.state.loading}
          addUrl={this.props.addUrl}
        />
      </div>
    );
  }

}

const mapStateToProps = (state) => {
  return ({
    auth: state.auth,
  })
}

export default connect(mapStateToProps, null)(Comments);
